### ISSUE TITLE
S2259 - False positive after is null check

### ISSUE THREAD

**Issue by snakefoot** [SonarSource/sonar-dotnet#6241]
### Description

S2259 - is null on at least one execution path

### Repro steps

```c#
        public static ReadOnlySingleBucketDictionary<TKey, IList<TValue>> BucketSort<TValue, TKey>(this IList<TValue> inputs, KeySelector<TValue, TKey> keySelector, IEqualityComparer<TKey> keyComparer)
        {
            Dictionary<TKey, IList<TValue>> buckets = null;
            TKey singleBucketKey = default(TKey);
            for (int i = 0; i < inputs.Count; i++)
            {
                TKey keyValue = keySelector(inputs[i]);

                if (i == 0)
                {
                    singleBucketKey = keyValue;
                }
                else if (buckets is null)
                {
                    if (!keyComparer.Equals(singleBucketKey, keyValue))
                    {
                        // Multiple buckets needed, allocate full dictionary
                        buckets = CreateBucketDictionaryWithValue(inputs, keyComparer, i, singleBucketKey, keyValue);
                    }
                }
                else
                {
                    if (!buckets.TryGetValue(keyValue, out var eventsInBucket)) // BUG: 'buckets' is null on at least one execution path
                    {
                        eventsInBucket = new List<TValue>();
                        buckets.Add(keyValue, eventsInBucket);
                    }
                    eventsInBucket.Add(inputs[i]);
                }
            }

            if (buckets is null)
                return new ReadOnlySingleBucketDictionary<TKey, IList<TValue>>(new KeyValuePair<TKey, IList<TValue>>(singleBucketKey, inputs), keyComparer);
            else
                return new ReadOnlySingleBucketDictionary<TKey, IList<TValue>>(buckets, keyComparer);
        }
```

### Expected behavior

Not reported as bug

### Actual behavior

Reported as bug

### Known workarounds

Rewrite code

### Related information

https://sonarcloud.io/project/issues?resolved=false&types=BUG&branch=dev&id=nlog2&open=AYP2NU0SZinl2h1ztXSa

**Comment by pavel-mikula-sonarsource** [comment]
Hi @snakefoot,

Thank you for reporting this case. I confirm it as False Positive. It looks pretty tricky - see the minimal reproducer in the PR above.

**Comment by pavel-mikula-sonarsource** [comment]
**Normal:** IsPattern with LocalReference and ConstantPattern. Where ConstantPattern has Literal value `null`.

**Generic:** IsPattern with LocalReference and ConstantPattern. Where ConstantPattern has a Conversion value, with Literal `null` value.
